### ISSUE TITLE
performance issues

### ISSUE THREAD

**Issue by agash75** [JacquesLucke/animation_nodes#1432]
Using Blender 2.82a
I have a7941b2b39c4491ebd8e5cd63a67aed6.zip installed (_2.1_py37_win64).
This is a version of 2.1.5 and it's working fine

In the latest verion, - 2.1.7 - doesn't make any difference if I grab it from AN website, or graphicall
the same node setup that executes in 1111ms, takes 78(!) seconds
On another scene, another node tree, the difference is 1200ms vs 28 seconds

One is a couple of objects instanced /duplicated 133 times, then randomized their location, rotation, and/or shape keys. The other scene is 133 curve objects that I join with the endpoints of another 133 curves by taking into account the interseting 1-5 objects, generating breakpoints and a spline.

By checking the execution time details, for the 1st scene, it looks liket it's the Object Transforms Input node that takes that much time to execute (it's inside of a loop).
For the 2nd, it's more interesting; I have 2 "Splines from Object" node in the loop, tha majority of 78sec is spread across them. When I turn the "Use World space" option off, the result becomes incorrect (expected), the execution time remains the same, but instead of the 2 "Splines from Object" nodes, it's already 1 Object Transforms Input (which is inside a loop inside the main loop) that becomes the bottleneck, taking 98% of total execution time (75 seconds!) alone.

**Comment by OmarEmaraDev** [comment]
Can you share a minimal blend file that replicates the issue? Also, can you try with the latest blender build?

**Comment by agash75** [comment]
Dear Omar,
The node tree is fairly complex, with probably not the most simplified (rather overcomplicated) logic behind, but I managed to achieve exactly what I wanted to (and, being my first AN ever, I'm happy with it).
[hangolokulcsok-es-szogek_03b_B28-ANTree_ref.zip](https://github.com/JacquesLucke/animation_nodes/files/4646178/hangolokulcsok-es-szogek_03b_B28-ANTree_ref.zip)
To replicate, simply change the instances in "NodeTree-string_posts" to 39.
Blender 2.82a with the above mentioned build of AN 2.1.5, it runs in 1103,2 ms, 
Blender 2.83 beta, AN 2.1.7 the same thing takes 23,4 sec

**Comment by OmarEmaraDev** [comment]
I don't see anything unusual here. The difference in execution time is probably due to the evaluations we started doing in recent versions before returning the transforms, which isn't a problem by itself.

I would say the main issue in your node tree is that you do a lot of unnecessary communications with Blender. Don't write data to Blender then read it again, do all the processing you need then output the data at the very end. For instance, here, you wrote the location to Blender then immediately read it again, just pass the location directly.

![20200518-223055](https://user-images.githubusercontent.com/25300994/82257108-c4222d00-9957-11ea-8d37-1e1b895db554.png)

**Comment by agash75** [comment]
As I said that was my first AN tree and not everything was (is) obvious.
I think first figured out how to arrange the pegs in the pattern I want, and only after that I randomized their transforms / shape keys.
I'm actually optimizing the scene, (like reducing the bevel objects resolution) and for that I need to make this 2.78/2.79 AN tree work in 2.82

But you're right Omar :) applying the transforms, then decompose the vector to do some other transformations is a complete nonsense. Despite of that, it was working and with 900ms execution time I didn't really think that would be an issue, until I ran into this horrible 25s execution time.
![Screenshot 2020-05-19 18 23 14](https://user-images.githubusercontent.com/65555796/82352179-e40c2c00-99fd-11ea-92e7-ee330252dc10.png)
Thx for the tip, by removing unnecessary node connections (ie. passing the raw vector data directly to the subprogram) the execution time instatly dropped, and now it's roughly the same in 2.82a/2.1.5 and 2.83/2.1.7 The latter is still slower but it's like 1200ms vs 1450ms or so, not 25s.
I'll check the other node too for similar issues :)

**Comment by 3DSinghVFX** [comment]
@agash75 To add, you can enable the _Measure Execution Times_ under the _Developers_ option to measure the times of each node to further debug the node-tree.
![Screenshot from 2020-05-20 11-34-33](https://user-images.githubusercontent.com/30294746/82411077-b1f6da80-9a8e-11ea-8610-45b9b078811f.png)

**Comment by agash75** [comment]
Hi,
Yes I have it. What I can't figure out, is where the slowdown really(!) comes from.
I have another node tree, and with the latest build from graphicakk, execution time is as horrible as 94 seconds! (shouldn't take more than a second)
One of the "splines from object" node is showing most of this time, 90s, while I have another of the same node (for another object group) where the info shows just a fraction of the above time.
Now, when I disonnect the input of the 1st, the stat shows that this 90s is coming from the 2nd node, If I disconnect both, then another node is shown as the culprit, while the total execution time remains - more or less - the same.
Is there a way to hunt down where these heavy calculations really(!) happen and what exactly it is? Is it possible to track/log AN-Blender communications?
Thank you

**Comment by OmarEmaraDev** [comment]
@agash75 This probably just means that Blender can avoid superfluous evaluations if nothing changes. So when you have two nodes that evaluates the spline, while both will return the splines, only one will do the evaluation. And if you remove it, the other will be the one doing the evaluation.

You are probably doing the same mistake we talked about. Why are you accessing the same object multiple times?

**Comment by Gerstmann-Bradley** [comment]
> Hi,
> Yes I have it. What I can't figure out, is where the slowdown really(!) comes from.
> I have another node tree, and with the latest build from graphicall, execution time is as horrible as 94 seconds! (shouldn't take more than a second)
> One of the "splines from object" node is showing most of this time, 90s, while I have another of the same node (for another object group) where the info shows just a fraction of the above time.
> Now, when I disonnect the input of the 1st, the stat shows that this 90s is coming from the 2nd node, If I disconnect both, then another node is shown as the culprit, while the total execution time remains - more or less - the same.
> Is there a way to hunt down where these heavy calculations really(!) happen and what exactly it is? Is it possible to track/log AN-Blender communications?
> Thank you

disclaimer: I am not a developer, just an AN user.
It's hard to know what you are actually talking about without a simplified node tree or a simplified file. (be aware of the word "simplified")
I have taken a look to the previous file you posted, it seems horrible and unfortunately I may not really understand what you are doing.
Just to mention for that specific file, it seems you want to instance multiple objects.
"object instancer" is not necessarily the only way you can go.
For example, AN has another option "transform mesh" node that can help replicate the mesh. An example can be seen in #1377 . Or, even blender itself has ability to "instance" object, and it may/may not combine with AN.

In general, I think many performance issues that you are facing may be rooted in the way you are noding. Maybe you are using wrong methods, maybe you are wasting nodes to calculate unnecessary staff, etc. These are my personal opinions.

**Comment by agash75** [comment]
@Gerstmann-Bradley That's when I used AN for the 1st time, that's why it's messy, as I moved forward I simply kept adding stuff without any optimizations :) And some silly nonsense too, as Omar pointed out :)
But the problem with the file posted is solved, now it executes in 1-2s which is fine for me, it doesn't need to run often, the generated collections are linked in another blend file.
What it basically does is its instancing multiple objects, placing them in a predefined pattern in calculated angle / spacing, and making some randomizations with Z rotation and/or shape keys. These are string posts and tuning keys of an instrument.

**Comment by agash75** [comment]
@OmarEmaraDev it's not the same object, it's 2 different group of curve objects, that's why I find the indicated execution times strange. 
Do you have any suggestion how I can check what's happening in the backgounrd to narrow down the problem?

**Comment by OmarEmaraDev** [comment]
@agash75 The mechanism I described could still apply even then. I can't give any suggestions if I am not familiar with the exact setup. If you could provide a minimal file to reproduce the the issue, that would be great.

**Comment by agash75** [comment]
@OmarEmaraDev Thank you, problem solved :)
You were right, it was my node structure, namely nested loops, which wasn't an issue with 2.79 and older version of AN where I put his together.
What was especially misleading is that AN showed that this huge execution time spike was caused by relatively simple nodes like "Objext Transforms Input" or "Mesh Object Input", but, even though I have multiple of those, AN showed always just one as the culprit. When I disconnected its input, execution time spike shown jumped to another one.
So I started by moving the position queries out of the loop and passing them as parameters, until I found one that was executed more than15k times... wth? Yes.
I use AN to connect 130 objects with another 130, while checking if there are any intersections with any of a 3rd group of objects (also about 130) and adding break points if necessary. 
I bet you figured out already, I was querying the vertex positions for all 130 on all 130 iterations, while I only need 2-7 of the 130, and, even worse (don't laugh) there was already another subprogram that returns the intersecting objects...
So I made a generic list with the vertex positions of the interseting objects, and filtering it with the subprogram's result (by index ID keys) brought the desired effect: execution time is down to1000ms :) 
Silly me. Thank you again for your time and patience :)
